cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -2815,43 +2814,62 @@ const char* soundex_utf8(gdv_int64 ctx, const char* in, gdv_int32 in_len,
   }
 
   // The soundex code is composed by one letter and three numbers
+  char* soundex = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(ctx, in_len));
   char* ret = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(ctx, 4));
-  if (ret == nullptr) {
+
+  if (soundex == nullptr || ret == nullptr) {
     gdv_fn_context_set_error_msg(ctx, "Could not allocate memory for output string");
     *out_len = 0;
     return "";
   }
 
   int si = 1;
+  int ret_len = 1;
   unsigned char c;
 
   int start_idx = 0;
   for (int i = 0; i < in_len; ++i) {
     if (isalpha(in[i]) > 0) {
+      // Retain the first letter
       ret[0] = toupper(in[i]);
       start_idx = i + 1;
       break;
     }
   }
 
-  for (int i = start_idx, l = in_len; i < l; i++) {
+  soundex[0] = '\0';
+  // Replace consonants with digits and special letters with 0
+  for (int i = start_idx; i < in_len; i++) {
     if (isalpha(in[i]) > 0) {
       c = toupper(in[i]) - 65;
-      if (mappings[c] != '0') {
-        if (mappings[c] != ret[si - 1]) {
-          ret[si] = mappings[c];
-          si++;
-        }
-
-        if (si > 3) break;
+      if (mappings[c] != soundex[si - 1]) {
+        soundex[si] = mappings[c];
+        si++;
       }
     }
   }
 
-  if (si <= 3) {
-    while (si <= 3) {
-      ret[si] = '0';
-      si++;
+  int i = 1;
+  // If the saved letter's digit is the same as the resulting first digit, remove the
+  // digit.
+  if (soundex[i] == mappings[ret[0] - 65]) {
```

Comment:
You have to handle the case where ret[0] is not initialised.. for e.g. the input is '123456' - since there are no alphabets, ret[0] will not be initialised

Similarly, you have to handle the case where soundex[1] is not set..

Reply:
Please change this to:
if (soundex[1] == mapp) {
 i = 2;
}

Reply:
> You have to handle the case where ret[0] is not initialised.. for e.g. the input is '123456' - since there are no alphabets, ret[0] will not be initialised
> 
> Similarly, you have to handle the case where soundex[1] is not set..

About the soundex[1];
If ret[0] is valid, we don't need handle with this case